File: objects/dsp.cxx
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.mikekohn.net/
 * License: GPL
 *
 * Copyright 2014 by Michael Kohn
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "JavaClass.h"
#include "compile.h"
#include "dsp.h"

#define CHECK_FUNC(funct) \
  if (strncmp(#funct, function, sizeof(#funct)-1) == 0) \
  { \
    return dsp_##funct(java_class, generator); \
  }

static int dsp_getA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_a();
}

static int dsp_getB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_b();
}

static int dsp_getUpperA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_upper_a();
}

static int dsp_getUpperB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_upper_b();
}

static int dsp_getLowerA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_lower_a();
}

static int dsp_getLowerB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_lower_b();
}

static int dsp_getRoundedA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_rounded_a();
}

static int dsp_getRoundedB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_get_rounded_b();
}

static int dsp_clearA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_clear_a();
}

static int dsp_clearB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_clear_b();
}

static int dsp_loadA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_load_a();
}

static int dsp_loadB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_load_b();
}

static int dsp_negA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_neg_a();
}

static int dsp_negB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_neg_b();
}

static int dsp_addABAndStoreInA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_add_ab_and_store_in_a();
}

static int dsp_addABAndStoreInB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_add_ab_and_store_in_b();
}

static int dsp_subABAndStoreInA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_sub_ab_and_store_in_a();
}

static int dsp_subBAAndStoreInB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_sub_ba_and_store_in_b();
}

static int dsp_addToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_add_to_a();
}

static int dsp_addToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_add_to_b();
}

static int dsp_squareToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_square_to_a();
}

static int dsp_squareToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_square_to_b();
}

static int dsp_mulToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_to_a();
}

static int dsp_mulToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_to_b();
}

#if 0
static int dsp_euclideanDistanceToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_to_a();
}

static int dsp_euclideanDistanceToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_to_b();
}
#endif

static int dsp_squareAndAddToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_square_and_add_to_a();
}

static int dsp_squareAndAddToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_square_and_add_to_b();
}

static int dsp_mulAndAddToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_and_add_to_a();
}

static int dsp_mulAndAddToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_and_add_to_b();
}

static int dsp_mulAndSubFromA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_and_sub_from_a();
}

static int dsp_mulAndSubFromB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_mul_and_sub_from_b();
}

#if 0
static int dsp_euclideanDistanceAndAddToA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_euclidean_distance_and_add_to_a();
}

static int dsp_euclideanDistanceAndAddToB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_euclidean_distance_and_add_to_b();
}
#endif

static int dsp_shiftA(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_shift_a();
}

static int dsp_shiftB(JavaClass *java_class, Generator *generator)
{
  return generator->dsp_shift_b();
}

int dsp(JavaClass *java_class, Generator *generator, char *function)
{
  CHECK_FUNC(getA)
  CHECK_FUNC(getB)
  CHECK_FUNC(getUpperA)
  CHECK_FUNC(getUpperB)
  CHECK_FUNC(getLowerA)
  CHECK_FUNC(getLowerB)
  CHECK_FUNC(getRoundedA)
  CHECK_FUNC(getRoundedB)
  CHECK_FUNC(clearA)
  CHECK_FUNC(clearB)
  CHECK_FUNC(loadA)
  CHECK_FUNC(loadB)
  CHECK_FUNC(negA)
  CHECK_FUNC(negB)
  CHECK_FUNC(addABAndStoreInA)
  CHECK_FUNC(addABAndStoreInB)
  CHECK_FUNC(subABAndStoreInA)
  CHECK_FUNC(subBAAndStoreInB)
  CHECK_FUNC(addToA)
  CHECK_FUNC(addToB)
  CHECK_FUNC(squareToA)
  CHECK_FUNC(squareToB)
  CHECK_FUNC(mulToA)
  CHECK_FUNC(mulToB)
  //CHECK_FUNC(euclideanDistanceToA)
  //CHECK_FUNC(euclideanDistanceToB)
  CHECK_FUNC(squareAndAddToA)
  CHECK_FUNC(squareAndAddToB)
  CHECK_FUNC(mulAndAddToA)
  CHECK_FUNC(mulAndAddToB)
  CHECK_FUNC(mulAndSubFromA)
  CHECK_FUNC(mulAndSubFromB)
  //CHECK_FUNC(euclideanDistanceAndAddToA)
  //CHECK_FUNC(euclideanDistanceAndAddToB)
  CHECK_FUNC(shiftA)
  CHECK_FUNC(shiftB)

  return -1;
}

File: java/net/mikekohn/java_grinder/DSP.java
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.naken.cc/
 * License: GPL
 *
 * Copyright 2014-2015 by Michael Kohn
 *
 */

package net.mikekohn.java_grinder;

abstract public class DSP
{
  static long A,B;

  protected DSP()
  {
  }

  /** Return value of A (only bits [31:16]) */
  public static int getA()
  {
    short n = (short)(A >> 16);
    return (int)n;
  }

  /** Return value of B (only bits [31:16]) */
  public static int getB()
  {
    short n = (short)(B >> 16);
    return (int)n;
  }

  /** Return value of A (only bits [38:23]) */
  public static int getUpperA()
  {
    short n = (short)(A >> 23);
    return (int)n;
  }

  /** Return value of B (only bits [38:23]) */
  public static int getUpperB()
  {
    short n = (short)(B >> 23);
    return (int)n;
  }

  /** Return value of A (only bits [23:8]) */
  public static int getLowerA()
  {
    short n = (short)(A >> 8);
    return (int)n;
  }

  /** Return value of B (only bits [23:8]) */
  public static int getLowerB()
  {
    short n = (short)(B >> 8);
    return (int)n;
  }

  /** Return value of A rounded */
  public static int getRoundedA()
  {
    return 0;
  }

  /** Return value of B rounded */
  public static int getRoundedB()
  {
    return 0;
  }

  /** Clear Accum A */
  public static void clearA() { A = 0; }

  /** Clear Accum B */
  public static void clearB() { B = 0; }

  /** Load Accum A */
  public static void loadA(int n) { A = n; }

  /** Load Accum B */
  public static void loadB(int n) { B = n; }

  /** Negate Accum A */
  public static void negA() { A = -A; }

  /** Negate Accum B */
  public static void negB() { B = -B; }

  /** Add Accum A with Accum B and store in Accum A */
  public static void addABAndStoreInA() { A = A + B; }

  /** Add Accum A with Accum B and store in Accum B */
  public static void addABAndStoreInB() { B = A + B; }

  /** Sub Accum A with Accum B (A-B) and store in Accum A */
  public static void subABAndStoreInA() { A = A - B; }

  /** Sub Accum A with Accum B (B-A) and store in Accum B */
  public static void subBAAndStoreInB() { B = B - A; }

  /** Add to Accum A */
  public static void addToA(int n) { A += n; }

  /** Add to Accum B */
  public static void addToB(int n) { B += n; }

  /** Square and store result in Accum A */
  public static void squareToA(int n) { A = n * n; }

  /** Square and store result in Accum B */
  public static void squareToB(int n) { B = n * n; }

  /** Multiply and store result in Accum A */
  public static void mulToA(int n, int m) { A = n * m; }

  /** Multiply and store result in Accum B */
  public static void mulToB(int n, int m) { B = n * m; }

  /** Euclidean distance to A */
  //public static void euclideanDistanceToA(int n, int m) { }

  /** Euclidean distance to B */
  //public static void euclideanDistanceToB(int n, int m) { }

  /** Square and add result in Accum A */
  public static void squareAndAddToA(int n) { A += n * n; }

  /** Square and add result in Accum B */
  public static void squareAndAddToB(int n) { B += n * n; }

  /** Multiply and add result in Accum A */
  public static void mulAndAddToA(int n, int m) { A += n * m; }

  /** Multiply and add result in Accum B */
  public static void mulAndAddToB(int n, int m) { B += n * m; }

  /** Multiply and substract result from Accum A */
  public static void mulAndSubFromA(int n, int m) { A -= n * m; }

  /** Multiply and substract result from Accum B */
  public static void mulAndSubFromB(int n, int m) { B -= n * m; }

  /** Euclidean distance added to A */
  //public static void euclideanDistanceAndAddToA(int n, int m) { }

  /** Euclidean distance added to B */
  //public static void euclideanDistanceAndAddToB(int n, int m) { }

  /** Arithmetic shift accumulator A (-16 to 16 bits only).
      Negative n means shift left, positive is shift right. */
  public static void shiftA(int n)
  {
    if (n > 0 && n <= 16) { A = A >> n; }
    else if (n < 0 && n >= -16) { A = A << (-n); }
  }

  /** Arithmetic shift accumulator B (-16 to 16 bits only) */
  public static void shiftB(int n)
  {
    if (n > 0 && n <= 16) { B = B >> n; }
    else if (n < 0 && n >= -16) { B = B << (-n); }
  }
}

